DatamartManagementService/DatamartManagementService.Infrastructure/Persistence/PayrollRepository.cs

```diff
@@ -0,0 +1,133 @@
+        Task<List<EmployeePayroll>> GetEmployeePayrollForCertainPeriodByEmployeeId(long id, DateTime startDate, DateTime endDate);
+        Task UpdateEmployeePayroll(EmployeePayroll updatePayroll);
+        Task UpdateEmployeePayrollDetail(EmployeePayrollDetail updatePayrollDetail);
+    }
+
+    public class PayrollRepository : IPayrollRepository
+    {
+        /// <summary>
+        /// Adding a pay period into payroll
+        /// </summary>
+        /// <param name="newPayroll"></param>
+        /// <returns></returns>
+        public async Task AddEmployeePayroll(EmployeePayroll newPayroll)
+        {
+            using var context = new RofDatamartContext();
+
+            context.EmployeePayroll.Add(newPayroll);
+
+            await context.SaveChangesAsync();
+        }
+
+        /// <summary>
+        /// Gets entire payroll history of EE
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns></returns>
+        public async Task<List<EmployeePayroll>> GetEmployeePayrollByEmployeeId(long id)
+        {
+            using var context = new RofDatamartContext();
+
+            var result = await context.EmployeePayroll.Where(ep => ep.EmployeeId == id).ToListAsync();
+
+            return result;
+        }
+
+        /// <summary>
+        /// Gets payroll history between a certain period of EE
+        /// </summary>
+        /// <param name="id"></param>
+        /// <param name="startDate"></param>
+        /// <param name="endDate"></param>
+        /// <returns></returns>
+        public async Task<List<EmployeePayroll>> GetEmployeePayrollForCertainPeriodByEmployeeId(long id, DateTime startDate, DateTime endDate)
+        {
+            using var context = new RofDatamartContext();
+
+            var employeePayroll = await context.EmployeePayroll.Where(ep => ep.EmployeeId == id).ToListAsync();
+
+            var result = new List<EmployeePayroll>();
+
+            foreach (var payroll in employeePayroll)
+            {
+                if (payroll.PayPeriodStartDate >= startDate && payroll.PayPeriodEndDate <= endDate)
+                {
+                    result.Add(payroll);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Updates payroll
+        /// </summary>
+        /// <param name="updatePayroll"></param>
+        /// <returns></returns>
+        public async Task UpdateEmployeePayroll(EmployeePayroll updatePayroll)
+        {
+            using var context = new RofDatamartContext();
+
+            context.Update(updatePayroll);
+
+            await context.SaveChangesAsync();
+        }
+
+        /// <summary>
+        /// Adds  detailed payroll info
+        /// </summary>
+        /// <param name="newPayrollDetail"></param>
+        /// <returns></returns>
+        public async Task AddEmployeePayrollDetail(EmployeePayrollDetail newPayrollDetail)

```

Comment:
this is to build a reporting database.
we will have a nightly job or something to insert payroll details which can come in bulk.
its better to do a bulk insert here and have it take in a list of payroll details